### PR TITLE
Async test cases

### PR DESCRIPTION
**Pull request by redshirtrob**
There's a lot going on here, but it's all mostly related.  The two main points:
 1.  Added nil checks on all completion handlers
 2.  Use the return value on `JSONObjectWithData:options:error:` to detect errors.

I also added three asynchronous test cases to cover these changes.

There's a bit of refactoring and formatting changes going on as well.  This is mostly to make the code style consistent in these files.

I will annotate the most interesting parts of the changes for code review purposes.

### CODE REVIEW

File: URXSearch/URXQuery.m [hunk 1]
```diff
@@ -12,25 +12,65 @@
...
++ (URXErrorType)urxErrorTypeForHTTPStatusCode:(NSInteger)statusCode;
+
+- (void)handleSearchResponse:(NSHTTPURLResponse *)response request:(NSURLRequest *)request data:(NSData *)data error:(NSError *)error successHandler:(void (^)(URXSearchResponse *))successHandler andFailureHandler:(void (^)(URXAPIError *))failureHandler;
 
 @end
 
 @implementation URXQuery
 
-- (URXAnd *) and:(URXQuery *)query {
++ (URXErrorType)urxErrorTypeForHTTPStatusCode:(NSInteger)statusCode {
```
Comment (redshirtrob) [inline URXSearch/URXQuery.m]: This is a refactor point.

File: URXSearch/URXQuery.m [hunk 2]
```diff
@@ -84,82 +125,56 @@ - (URXSearchResponse *) searchSynchronously {
...
-    
-    URXAPIError *(^makeError)(URXErrorType)= ^URXAPIError *(URXErrorType e) {
-        return [[URXAPIError alloc] initWithErrorType:e request:request response:response error:error data:data];
-    };
-    
-    //If the HTTP Request gave us an error or we didn't get any data, fail
-    if (error != nil || [data length] == 0){
-        failureHandler(makeError(URXNetworkConnectionError));
-        return;
+- (void)handleSearchResponse:(NSHTTPURLResponse *)response request:(NSURLRequest *)request data:(NSData *)data error:(NSError *)error successHandler:(void (^)(URXSearchResponse *))successHandler andFailureHandler:(void (^)(URXAPIError *))failureHandler {
```
Comment (redshirtrob) [inline URXSearch/URXQuery.m]: The main changes are here.  I use the new method for mapping HTTP status codes to URX Error codes and add nil checks on the completion handlers.

File: URXSearch/URXResolutionRequest.m [hunk 3]
```diff
@@ -13,24 +13,65 @@ @interface URXResolutionRequest() {
...
 @end
 
 @implementation URXResolutionRequest
 
--(void) resolveAsynchronouslyWithSuccessHandler:(void (^)(URXResolutionResponse *))successHandler andFailureHandler:(void (^)(URXAPIError *))failureHandler{
++ (instancetype)requestFromUrl:(NSString *)urlString {
+    return [[self alloc] initWithUrlString:urlString];
+}
+
++ (URXErrorType)urxErrorTypeForHTTPStatusCode:(NSInteger)statusCode {
```
Comment (redshirtrob) [inline URXSearch/URXResolutionRequest.m]: Same as the `URXQuery` change.

File: URXSearch/URXResolutionRequest.m [hunk 4]
```diff
@@ -66,87 +107,65 @@ -(URXResolutionResponse *) resolveSynchronously {
     return resolutionResponse;
 }
 
--(void) handleResolutionResponse:(NSHTTPURLResponse *)response request:(NSURLRequest *)request data:(NSData *)data error:(NSError *)error successHandler:(void (^)(URXResolutionResponse *))successHandler andFailureHandler:(void (^)(URXAPIError *))failureHandler {
-    URXAPIError *(^makeError)(URXErrorType)= ^URXAPIError *(URXErrorType e) {
-        return [[URXAPIError alloc] initWithErrorType:e request:request response:response error:error data:data];
-    };
+- (void)handleResolutionResponse:(NSHTTPURLResponse *)response request:(NSURLRequest *)request data:(NSData *)data error:(NSError *)error successHandler:(void (^)(URXResolutionResponse *))successHandler andFailureHandler:(void (^)(URXAPIError *))failureHandler {
```
Comment (redshirtrob) [inline URXSearch/URXResolutionRequest.m]: Same as the `URXQuery` change.

File: URXSearch/URXQuery.m (hunk 1 above)
Comment (jlturner) [inline URXSearch/URXQuery.m]: :+1:

File: URXSearch/URXQuery.m [hunk 5]
```diff
@@ -84,82 +125,56 @@ - (URXSearchResponse *) searchSynchronously {
...
+     *   than the error parameter for not nil.  Logically,
+     *   (1)  return value == nil -> Error
+     *   (2)  return value == nil -> `error` will contain NSError object
+     *
+     *   The docs do not guarantee that the converse (3) of (2) is true:
+     *   (3)  `error` != nil -> Error  (Not necessarily true)
+     *
+     *   In my experience statement (3) has always been true.  But, I do not want to rely on it.
+     */
+    NSDictionary *responseJSON = [NSJSONSerialization JSONObjectWithData:data options:0 error:&error];
```
Comment (jlturner) [inline URXSearch/URXQuery.m]: Makes sense, this has never been an issue for me in practice (I actually always checked if `error` was `nil`, but after thinking over your argument, I think checking the data returned makes more sense overall).

File: URXSearchTests/URXIntegrationTests.m [hunk 6]
```diff
@@ -29,4 +31,87 @@ -(void)testIntegration {
...
+            XCTFail(@"Integration test couldn't resolve search request for query \"%@\". Error: %@", URXIntegrationQueryValue, response.error.errorMessage);
+            [expectation fulfill];
+        }];
+    [self waitForExpectationsWithTimeout:60.0 handler:nil];
+}
+
+/*
+ * The tests below test for SDK crashes on nil completion blocks.  We don't have a very good way to detect
+ * that a particular HTTP request is done so we just blindly wait some amount of time and assume
+ * the test succeeded if the test didn't crash.  There's no failure condition, just crashes.
```
Comment (jlturner) [inline URXSearchTests/URXIntegrationTests.m]: Yep. Timeout is pretty much all we can do here.

**Comment by jlturner** [comment]
LGTM :+1:

**Pull request merged**